x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,728 @@
+- **Bulk Actions** - `POST /rules/_bulk_action`: 
+    - This endpoint includes many types of actions, and only a subset of them should perform a migration.
+    - This endpoint also includes a `dry_run` mode that is executed to evaluate preconditions and warn the user before executing the actual request. No migration logic should take place for dry run requests, i.e when `dry_run=true`.
+    - Actions that should perform migration logic:
+        - Edit:
+            - endpoint should migrate rule while performing migration. We can take advantage of the `ruleParamsModifier` to carry out the migration, regardless of the type of edit that is being performed.
+    - Actions that should **not** perform migration logic:
+        - Enable
+        - Disable
+        - Delete
+        - Export:
+            - export endpoint logic does not write/modify rules, only reads them and makes them available for download. Adding new writing logic could have a big impact on its performance.
+            - we will now allow the export of prebuilt rules
+            - export endpoint should therefore only export the rule in their current state (migrated or non-migrated), and the import endpoint will take care of the migration logic
+        - Duplicate:
+            - this action should not migrate the original rule being duplicated, as it would violate the principle of least surprise
+            - the new rule being created should have the new schema
+            - we should only update the duplication logic to account for possibility that the new `prebuilt` and `customized` fields already exist in the original rule
+            - per definition: if a prebuilt rule is duplicated, we need to set the new rule's `prebuilt` value to `false` (and keep the current logic that also sets `immutable` to `false`).
+
+**NOTE:** Endpoints that depend on the Rules Client's methods need to modify the rules' params (prebuilt and customized) before the Rule Client methods call the`validateRuleTypeParams`, because that util validates the passed parameters using the validators defined in `x-pack/plugins/security_solution/server/lib/detection_engine/rule_types`. Since those are an intersection between `BaseRulesParams` and the specific rule type's fields, `prebuilt` and `customized` are non-optional, which would cause the util to reject the validation of the params for non-migrated rules.
+
+### Other Rule Management endpoints that will need to be adapted to new schema, but will not need migration logic:
+
+- **Create Rules** - `POST /rules`:
+    - If the request includes either the `immutable` and `prebuilt` fields, ignore both parameters and set both fields to `false`. This is analogous to the current logic: we don't want users to be able to create prebuilt rules, only install them and customize them, and always be able to compare them to a `security-rule` asset.
```

Comment:
Currently, we don't support the `immutable` field in any of the endpoints' request parameters. We shouldn't support the `prebuilt` and `customized` fields either, because their values should be controlled by the app on the server side, and never controlled by users.

Specifically, this endpoint should be able to create _only_ rules with:

```
immutable: false,
prebuilt: false,
customized: false
```

I guess this is what you meant by saying "ignore both parameters and set both fields to `false`", but I think it should be explicitly called out that all the three fields shouldn't be supported in request parameters of all our endpoints, except maybe the rules import endpoint.

Reply:
Thanks, I added and clarified this.